### ISSUE TITLE
Design Panel: Changes do not save when clicking back into main story editor area

### ISSUE THREAD

**Issue by mxbclang** [GoogleForCreators/web-stories-wp#9000]
## Bug Description

A [forum user reported](https://wordpress.org/support/topic/new-bugs-image-size-validation-more/) that when a user adds or changes text in the Accessibility field for an image and then clicks back into the main editor area, the Accessibility text is not saved.

## Expected Behaviour

Accessibility text should be saved by default after the user clicks back into the main editor area.

## Steps to Reproduce

1. Click on an image in the editor
2. In the right sidebar, enter text in the Accessibility field
3. Click anywhere in the main editor area
4. Click on the original image and see that Accessibility text was not added/updated

## Screenshots

Video: https://jmp.sh/FwLr6y5

## Additional Context

- Plugin Version: v1.110
- WordPress Version: v5.8.1
- Operating System: macOS Big Sur
- Browser: Chrome v93

**Comment by swissspidy** [comment]
I think this is true for all changes in the design panel. Can you confirm that?

cc @barklund

**Comment by mxbclang** [comment]
@swissspidy The user originally reported it for Accessibility and Attachment, but I can't replicate with Attachment – things are saving for me there.

**Comment by miina** [comment]
Tested a few fields and this seems to be like this for most of the fields, e.g. when changing a value of Width of an element (without leaving the input) and then selecting the Page, the value does not get assigned. 

It happens not when clicking anywhere in the main editor but when changing selection before blurring, the value doesn't get saved then.

It does not happen for all fields though, e.g. Page Attachment fields and Link fields save the value anyway. But yes, this is the default behaviour for most of the fields.

**Comment by swissspidy** [comment]
Thanks for confirming @miina!

Is this fixable?

**Comment by miina** [comment]
Most likely fixable but not fully sure what would be the effort of this and if this would mean sacrificing any other features.

The issue here is that even though the fields have similar logic in the editor, then in the design system the elements behave differently. E.g. the numeric input is not actually passing `onChange` to the input directly but is instead saving the value locally on change, no matter what the editor has for the `onChange` logic. The numeric input was implemented in https://github.com/google/web-stories-wp/pull/6419
It also includes things like restoring the original value with `Esc` etc. 

That's the same for TextArea (having internal `onChange` handler) which is used by the accessibility fields and was reported in this GH issue.

The link description and other fields that update as expected are using `Input` field from the design system which does not have inner custom handling of `onChange`.

For making it work for all the inputs, it would require reworking the inputs so that perhaps the design system would be purely presentational and not have its own logic for `onChange`. It would need a bit of further investigation to understand the complexity/possibility of maintaining the current features as well while removing the `onChange` internal handling.

Perhaps @samwhale remembers the requirements for needing internal `onChange` handlers. I can see the following:
- Saving the value only `onBlur` (to enforce not calling `onChange` with each character change)
- Allowing to restore original value by using `Esc`

Were there any others that depended on this?

Related: https://github.com/google/web-stories-wp/issues/6690

**Comment by samwhale** [comment]
👋 hello! Yes, these components are all controlled components except for the `NumericInput`.

>Perhaps @samwhale remembers the requirements for needing internal onChange handlers. I can see the following:

>- Saving the value only onBlur (to enforce not calling onChange with each character change)
>- Allowing to restore original value by using Esc

The `NumericInput` is a little different - we didn't want to blow up the existing functionality in the app so this input tracks its own internal state to allow for reverting changes.

The `NumericInput` calls the `onChange` handler that's passed in when the input is blurred:
https://github.com/google/web-stories-wp/blob/37835ce9a2d110e71c7ddef1e40415efa688f04c/packages/design-system/src/components/input/useNumericInput.js#L49-L80

@miina is totally right! Reverting the state when pressing `Esc` and not wanting to push every update on every key stroke led us to using internal change handlers. The ⬆️  and ⬇️ keys also change the value, but that's it.

---

>For making it work for all the inputs, it would require reworking the inputs so that perhaps the design system would be purely presentational and not have its own logic for `onChange`. It would need a bit of further investigation to understand the complexity/possibility of maintaining the current features as well while removing the `onChange` internal handling.

I think you're totally right about the reason this doesn't work. It's because we only save values on blur rather than on change BUT switching the selected element doesn't bubble up an event that would trigger a blur.

Ideally we'd save values with the `onChange` handler rather than blurring. We could even debounce network requests or saves so that we don't save on every single key stroke. Then we shouldn't run into this problem.

**Comment by miina** [comment]
cc @barklund For prioritization
